### PR TITLE
tree: Basic schema editing support

### PR DESCRIPTION
**Pull request by taylorsw04**
## Description

This PR introduces basic support for schema editing. It makes schema edits a first-class concept, adding support for it in the SharedTree op type and handling it in rebasing.

## Breaking Changes

This moves schema editing off of the stored schema and onto the view. It also breaks the persisted format by changing the top-level op format.

### CODE REVIEW

File: experimental/dds/tree2/src/test/shared-tree/summary.bench.ts
Comment (taylorsw04) [inline experimental/dds/tree2/src/test/shared-tree/summary.bench.ts:28]: This increase is necessary as the new schema op contains both the new schema and the previous schema. This may warrant the proposed reduction in schema index size.

Comment (taylorsw04) [reply experimental/dds/tree2/src/test/shared-tree/summary.bench.ts:28]: Alternatively, it would be fairly straightforward to make the schema change delta-based as the in-memory format is a btree that can efficiently compute deltas.

File: experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts
Comment (taylorsw04) [inline experimental/dds/tree2/src/test/shared-tree-core/sharedTreeCore.spec.ts:284]: This line is required to make this test actually test what it claims; however, without this line, it fails in delta application. This is a bug and should be investigated.

File: experimental/dds/tree2/src/feature-libraries/default-schema/defaultEditBuilder.ts
Comment (taylorsw04) [inline experimental/dds/tree2/src/feature-libraries/default-schema/defaultEditBuilder.ts:166]: This method was unused and not part of an interface.

File: experimental/dds/tree2/src/shared-tree/schematizedTree.ts
Comment (taylorsw04) [inline experimental/dds/tree2/src/shared-tree/schematizedTree.ts:172]: Can't think of a cleaner way to do this.

File: experimental/dds/tree2/src/feature-libraries/modular-schema/modularChangeCodecs.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/modular-schema/modularChangeCodecs.ts:47]: What is the motivation to remove codec versioning for this?

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts:122]:
```suggestion
		 * become out of schema if a schema changes to restrict root types to just A is concurrent with a data change that sets it to a B.
```

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts:128]: Any reason this is re-declared for every call of this method?

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts:117]: This comment ends up being associated with `hasSchemaChange` which seems a little arbitrary.
Perhaps a `//`-based comment would make more sense here.

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts:131]: Why conflict when `over.change.changes.length === 0`?

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts:130]: This may be personal preference, but when `change.changes.length === 0`, I would find it less confusing to return `change`. Especially since there's a coment in the scope below talking about conflict which this isn't.

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts:135]: I think this would be a better place for the big comment above.

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeCodecs.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeCodecs.ts:18]: Abram used \src\codec\discriminatedUnions.ts for representing discriminated enums in the wire format for `MarkEffect`. You may want to check with him if it makes sense to do the same here.

File: experimental/dds/tree2/src/shared-tree/sharedTreeEditBuilder.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree/sharedTreeEditBuilder.ts:21]: I see you opted against the design we had discussed last where this would have exposed a `schema` getter that returned an object with all of schema's editing methods (currenly only `set` but I immagine that could change). I'm curious why.

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeCodecs.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeCodecs.ts:29]: Give the drive to keep our encoded format small, I have to ask: do we really benefit from naming this `encodedChanges` instead of `changes`?

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts
Comment (taylorsw04) [reply experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts:131]: Whoops, good catch!

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts
Comment (CraigMacomber) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts:31]:
```suggestion
	public static readonly emptyChange: SharedTreeChange = {
```

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeTypes.ts
Comment (CraigMacomber) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeTypes.ts:19]:
```suggestion
		| { readonly type: "data"; readonly innerChange: ModularChangeset }
		| { readonly type: "schema"; readonly innerChange: SchemaChange }
```

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeTypes.ts
Comment (CraigMacomber) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeTypes.ts:15]: It would be nice to document why this restriction exists. It sounds like an optimization: is there a reason somone can't have two changes in a row in a SharedTreeChange, for example as an intermediate version before coalescing them or as part of a deoptimized reference version?

Also technically typescript likely can express that constraint. I wouldn't recommend doing it (its a real mess of recursive types) but I'm think it might be doable.

File: experimental/dds/tree2/src/feature-libraries/modular-schema/modularChangeCodecs.ts
Comment (taylorsw04) [reply experimental/dds/tree2/src/feature-libraries/modular-schema/modularChangeCodecs.ts]: Discussed offline. Reverted!

File: experimental/dds/tree2/src/shared-tree/sharedTreeEditBuilder.ts
Comment (taylorsw04) [reply experimental/dds/tree2/src/shared-tree/sharedTreeEditBuilder.ts:21]: Done!

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeTypes.ts
Comment (taylorsw04) [reply experimental/dds/tree2/src/shared-tree/sharedTreeChangeTypes.ts:19]: Added a comment.

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeCodecs.ts
Comment (CraigMacomber) [inline experimental/dds/tree2/src/shared-tree/sharedTreeChangeCodecs.ts:26]: this looses any possible codec options other than the validator. I think it would be better to not destructure and recreate the options like this.

File: experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json:24]: the lack of versions at lower levels of indent in this file is a bit concerning.
What version do you bump when making "old" not included for example? I don't think that needs to block this PR, but is something we really need to sort out.

File: experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json:63]: the new and old schema here look the same. That seems wrong to me. Shouldn't old be the default empty document schema? Either that of this change shouldn't have been done in the first place.

File: experimental/dds/tree2/src/test/snapshots/files/insert-and-delete-tree-0-final.json
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/snapshots/files/insert-and-delete-tree-0-final.json:24]: do we actually need to add 3 levels of nesting here? That seems like a bit much.

Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/snapshots/files/insert-and-delete-tree-0-final.json:24]: I'm fine with working on that as a followup, but I think we might want to track it, and I think we might have some design patterns that lead to more nesting overhead than we really want in our persisted data

File: experimental/dds/tree2/src/shared-tree/sharedTreeChangeCodecs.ts
Comment (taylorsw04) [reply experimental/dds/tree2/src/shared-tree/sharedTreeChangeCodecs.ts:26]: Done.

File: experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json
Comment (taylorsw04) [reply experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json:63]: Fixed! Nice catch. As discussed, I will add a test for this issue.

File: experimental/dds/tree2/src/test/snapshots/files/insert-and-delete-tree-0-final.json
Comment (taylorsw04) [reply experimental/dds/tree2/src/test/snapshots/files/insert-and-delete-tree-0-final.json:24]: Will address before merging.

File: experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json
Comment (taylorsw04) [reply experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json:24]: Will file an item before merging.

File: experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json
Comment (taylorsw04) [reply experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json:63]: Done.

File: experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json
Comment (taylorsw04) [reply experimental/dds/tree2/src/test/snapshots/files/empty-root-final.json:24]: Item filed.

File: experimental/dds/tree2/src/test/snapshots/files/insert-and-delete-tree-0-final.json
Comment (taylorsw04) [reply experimental/dds/tree2/src/test/snapshots/files/insert-and-delete-tree-0-final.json]: Done!

**Review by yann-achard-MS — APPROVED** [review]

**Pull request merged**
